packages/adapters/subgraph/src/lib/helpers/parse.ts

```diff
@@ -1,69 +1,148 @@
 import { XTransfer } from "@connext/nxtp-utils";
 import { BigNumber } from "ethers";
 
-export const xtransfer = (subgEntity: any): XTransfer => {
+// Used for sanity checking: must have these fields to be identified as a Transfer entity.
+export const TRANSFER_ENTITY_REQUIREMENTS = [
+  "transferId",
+  "nonce",
+  "to",
+  "callData",
+  "originDomain",
+  "destinationDomain",
+];
+
+export const transferEntitySanityCheck = (entity: any) => {
+  if (!entity) {
+    throw new Error("Subgraph entity parser: Transfer entity is `undefined`.");
+  }
+  for (const field of TRANSFER_ENTITY_REQUIREMENTS) {
+    if (!entity[field]) {
+      throw new Error(`Subgraph entity parser: Transfer entity missing required field: ${field}`);
+    }
+  }
+};
+
+export const originTransfer = (entity: any): XTransfer => {
+  transferEntitySanityCheck(entity);
+  if (entity.status || entity.originSender || entity.routers) {
+    // Wrong transfer type. This is a destination transfer entity!
+    throw new Error("Subgraph `OriginTransfer` entity parser: Transfer entity is a destination transfer entity.");
+  }
+  return {
+    // Meta Data
+    idx: entity.idx ? entity.idx : undefined,
+    transferId: entity.transferId,
+    nonce: BigNumber.from(entity.nonce ?? "0").toNumber(),
+
+    // Call Params
+    to: entity.to,
+    callData: entity.callData,
+
+    // Origin Info
+    origin: {
+      domain: entity.originDomain,
+
+      // Assets
+      assets: {
+        transactingAsset: entity.transactingAsset,
+        transactingAmount: entity.transactingAmount,
+        bridgedAsset: entity.bridgedAsset,
+        bridgedAmount: entity.bridgedAmount,
+      },
+
+      // XCall
+      xcall: {
+        // Event Data
+        relayerFee: entity.relayerFee,
+        // Transaction Data
+        caller: entity.xcalledCaller,
+        transactionHash: entity.xcalledTransactionHash,
+        timestamp: BigNumber.from(entity.xcalledTimestamp ?? "0").toNumber(),
+        gasPrice: entity.xcalledGasPrice,
+        gasLimit: entity.xcalledGasLimit,
+        blockNumber: BigNumber.from(entity.xcalledBlockNumber ?? "0").toNumber(),
+      },
+    },
+
+    // Destination Info
+    destination: {
+      domain: entity.destinationDomain,
+
+      status: undefined,
+      assets: undefined,
+      execute: undefined,
+      reconcile: undefined,
+    },
+  };
+};
+
+export const destinationTransfer = (entity: any): XTransfer => {
+  transferEntitySanityCheck(entity);
```

Comment:
it looks like it will be failing in the sanitationCheck because the entity can be executed/reconciled. we don't have `nonce`, `to` and `destinationDomain` for reconciled transfers.   how do you think?

Reply:
oh, really? if so, excellent catch, let me see